pkg/proxy/proxy/transparent.go

```diff
@@ -117,60 +118,104 @@ func NewTransparentProxyService(p *goproxy.ProxyHttpServer, ca *tls.Certificate,
 
 // ProxyHTTP serves an endpoint that transparently redirects HTTP connections to the proxy server.
 // This endpoint also explicitly (i.e. non-transparently) proxies HTTP and TLS connections.
-func (t TransparentProxyService) ProxyHTTP(addr string) {
+func (t TransparentProxyService) ProxyHTTP(addr string) func(context.Context) error {
 	ln, err := net.Listen("tcp", addr)
 	if err != nil {
 		log.Fatalf("Error listening for http connections - %v", err)
 	}
-	if err := http.Serve(ln, t.Proxy); err != nil {
-		log.Fatalln(err)
+	server := &http.Server{
+		Addr:    addr,
+		Handler: t.Proxy,
 	}
+	go func() {
+		if err := server.Serve(ln); err != nil {
+			log.Println(err)
+		}
+	}()
+	return func(ctx context.Context) error { return server.Shutdown(ctx) }
 }
 
 // ProxyTLS serves an endpoint that transparently redirects TLS connections to the proxy server.
-func (t TransparentProxyService) ProxyTLS(addr string) {
+func (t TransparentProxyService) ProxyTLS(addr string) func(context.Context) error {
 	ln, err := net.Listen("tcp", addr)
 	if err != nil {
 		log.Fatalf("Error listening for https connections - %v", err)
 	}
-	for {
-		log.Println("Awaiting TLS connection")
-		c, err := ln.Accept()
-		if err != nil {
-			log.Printf("Error accepting new connection - %v", err)
-			continue
-		}
-		go func(c net.Conn) {
-			log.Println("Connection from ", c.RemoteAddr())
-			conn, hello, err := handshake.PeekClientHello(c)
-			if err != nil {
-				log.Printf("Error accepting new connection - %v", err)
+	var inflightRequests sync.WaitGroup
+	t.Proxy.OnResponse().DoFunc(func(resp *http.Response, ctx *goproxy.ProxyCtx) *http.Response {
+		defer inflightRequests.Done()
+		return resp
+	})
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-done:
```

Comment:
I think in the steady state, this thread will block at line 157's `ln.Accept()` so you could detect net.ErrClosed as the error return instead to terminate the thread. That would remove the select and the extra channel creation which should simplify things